gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/activity/impl/ProcessWorkUnitImpl.java

```diff
@@ -164,46 +160,58 @@ public void onTaskCommitCompletion(Task task) {
+    if (fileSourcePaths.isEmpty()) {
+      // TODO - describe WorkUnits other than `CopyableFile`
+      return Optional.empty();
+    } else {
+      return Optional.of(getSourcePathsToLog(fileSourcePaths, jobState)).map(pathsToLog ->
+          "for copying source files: "
+              + (pathsToLog.size() == workUnits.size() ? "" : ("**first " + pathsToLog.size() + " only** "))
+              + pathsToLog
+      );
+    }
   }
 
-  protected Optional<CopyableFile> getOptCopyableFile(TaskState taskState) {
-    return getOptCopyableFile(taskState, "taskState '" + taskState.getTaskId() + "'");
+  protected static Optional<String> getOptCopyableFileSourcePathDesc(WorkUnit workUnit, String workUnitPath) {
+    return getOptWorkUnitCopyEntityClass(workUnit, workUnitPath).flatMap(copyEntityClass ->
+        getOptCopyableFile(copyEntityClass, workUnit).map(copyableFile ->
+            copyableFile.getOrigin().getPath().toString()));
   }
 
-  protected Optional<CopyableFile> getOptCopyableFile(State state, String logDesc) {
-    return getOptCopyEntityClass(state, logDesc).flatMap(copyEntityClass -> {
-      log.debug("(state) {} got (copyEntity) {}", state.getClass().getName(), copyEntityClass.getName());
-      if (CopyableFile.class.isAssignableFrom(copyEntityClass)) {
-        String serialization = state.getProp(CopySource.SERIALIZED_COPYABLE_FILE);
-        if (serialization != null) {
-          return Optional.of((CopyableFile) CopyEntity.deserialize(serialization));
-        }
+  protected static Optional<CopyableFile> getOptCopyableFile(TaskState taskState) {
+    return getOptTaskStateCopyEntityClass(taskState).flatMap(copyEntityClass ->
+        getOptCopyableFile(copyEntityClass, taskState));
+  }
+
+  protected static Optional<CopyableFile> getOptCopyableFile(Class<?> copyEntityClass, State state) {
+    log.debug("(state) {} got (copyEntity) {}", state.getClass().getName(), copyEntityClass.getName());
+    if (CopyableFile.class.isAssignableFrom(copyEntityClass)) {
+      String serialization = state.getProp(CopySource.SERIALIZED_COPYABLE_FILE);
+      if (serialization != null) {
+        return Optional.of((CopyableFile) CopyEntity.deserialize(serialization));
       }
-      return Optional.empty();
-    });
+    }
+    return Optional.empty();
+  }
+
+  protected static Optional<Class<?>> getOptWorkUnitCopyEntityClass(WorkUnit workUnit, String workUnitPath) {
+    return getOptCopyEntityClass(workUnit, "workUnit '" + workUnitPath + "'");
   }
 
-  protected Optional<CopyableFile> getOptFirstCopyableFile(List<WorkUnit> workUnits, String workUnitPath) {
-    return Optional.of(workUnits).filter(wus -> wus.size() > 0).flatMap(wus ->
-      getOptCopyableFile(wus.get(0), "workUnit '" + workUnitPath + "'")
-    );
+  protected static Optional<Class<?>> getOptTaskStateCopyEntityClass(TaskState taskState) {
+    return getOptCopyEntityClass(taskState, "taskState '" + taskState.getTaskId() + "'");
   }
```

Comment:
Wondering if we need to get copyEntityClass at a workunit level explicitly here since these functions imply that you can have multiple at a taskstate and workunit level, afaik these are tied to the source class and there can only be one source class per Gobblin job?

Reply:
is your suggestion to not consult each individual WU, but instead just sample one and acquire the info from that, presuming all WUs to be homogenous?

semantically you're probably right about current practice of a single source producing homogenous WUs (not 100% certain, but I suspect it's true).  I also have no any reason to believe that would change any time soon.

more as a general rule, my pref is to avoid spreading presumptions around various unrelated parts of code, but instead to code defensively.  e.g. I can *imagine* blending WUs of different types (even if we don't do that today).  if that were to ever come about, it would break such a presumption here.  hence the defensive approach of not using prior, outside knowledge to build presumptions.

so that's my pref, but what do you think?

Reply:
I understand it from a defensive standpoint, yes that was the idea I had in mind.
The reason why I suggested it is generally because (in my limited opinion) the framework should be opinionated around certain ideas, and right now a general concept in Gobblin is that workunits are generally homogenous at a WU level (hence why it may make sense to not conflate job state with task state/WU properties but currently that wasn't strongly enforced and led to this situation). If we have stronger presumptions that we are explicit about the code can be easier to reason about rather than learning about mismatches in behaviors after the implementation (which is what we're sort of seeing in the CommitActivity and Task states, though that is unrelated to this PR).

Reply:
I definitely agree--great system design philosophy!  basically define core requirements that translate into predictable system invariants, against which we reduce the conceptual surface area for maintainers (forever and after!).  clearly a big potential win!  the trick is to identify what restrictions we'll be comfortable accepting.

in this case, I DO expect we'd all accept the "single source, homogenous WUs" restriction.

the other element is enforcement: to preclude deviations, impl code generally needs to actively look for and reject violations.  it's still defensive coding (never let that go!), but just situating it earlier in the system flow (gating entry, if you will, rather than when manipulating data that's already entered the system).

since I'm not aware of where such enforcement lives in this case, I still prefer this defensive coding at point of use, but if you feel strongly, I'm open to reconsidering.